AnkiDroid/src/main/java/com/ichi2/libanki/importer/Anki2Importer.java

```diff
@@ -226,48 +238,55 @@ private void _importNotes() {
+                        id += 999;
                     }
-                    existing.add((Long) note[0]);
+                    existing.add(id);
                     // bump usn
-                    note[4] = usn;
                     // update media references in case of dupes
-                    note[6] = _mungeMedia((Long) note[MID], (String) note[6]);
-                    add.add(note);
-                    dirty.add((Long) note[0]);
+                    flds = _mungeMedia(mid, flds);
+                    add.add(new Object[]{id, guid, mid, mod, usn, tags, flds, sfld, csum, flag, data});
+                    dirty.add(id);
                     // note we have the added guid
-                    mNotes.put((String) note[GUID], new Object[]{note[0], note[3], note[MID]});
+                    mNotes.put(guid, new NoteTriple(id, mod, mid));
                 } else {
                     // a duplicate or changed schema - safe to update?
                     dupes += 1;
                     if (mAllowUpdate) {
-                        Object[] n = mNotes.get(note[GUID]);
-                        //todo: oldNid could be Long instead of long.
-                        long oldNid = (Long) n[0];
-                        long oldMod = (Long) n[1];
-                        @NonNull Long oldMid = (Long) n[2];
+                        NoteTriple n = mNotes.get(guid);
+                        long oldNid = n.mNid;
+                        long oldMod = n.mMod;
+                        long oldMid = n.mMid;
                         // will update if incoming note more recent
-                        if (oldMod < (Long) note[MOD]) {
+                        if (oldMod < mod) {
                             // safe if note types identical
-                            if (Utils.equals(oldMid, note[MID])) {
+                            if (Utils.equals(oldMid, mid)) {
                                 // incoming note should use existing id
-                                note[0] = oldNid;
-                                note[4] = usn;
-                                note[6] = _mungeMedia((Long) note[MID], (String) note[6]);
-                                update.add(note);
-                                dirty.add((Long) note[0]);
+                                id = oldNid;
+                                flds = _mungeMedia(mid, flds);
+                                update.add(new Object[]{id, guid, mid, mod, usn, tags, flds, sfld, csum, flag, data});
+                                dirty.add(id);
                             } else {
                                 dupesIgnored.add(String.format("%s: %s",
                                         mCol.getModels().get(oldMid).getString("name"),
-                                        ((String) note[6]).replace("\u001f", ",")));
-                                mIgnoredGuids.put((String) note[GUID], true);
+                                        flds.replace("\u001f", ",")));
```

Comment:
Use a char here - should be much faster (needs testing)
```suggestion
                                        flds.replace('\u001f', ',')));
```

Reply:
https://github.com/ankidroid/Anki-Android/pull/7598